nexus/social_media/services.py

```diff
@@ -0,0 +1,173 @@
+                                                                headers,
+                                                                post,
+                                                                api_url_base,
+                                                                linkedin)
+
+        if response_upload_image.status_code == status.HTTP_201_CREATED:
+            shareContent = post_data.get('specificContent')\
+                                    .get('com.linkedin.ugc.ShareContent')
+            image_media = [{
+                "status": "READY",
+                "description": {
+                    "text": ""
+                },
+                "media": asset,
+                "title": {
+                    "text": ""
+                }
+            }]
+            shareContent.update(shareMediaCategory='IMAGE',
+                                media=image_media)
+            image_response = requests.post(api_url_ugc, headers=headers,
+                                           json=post_data)
+            return image_response
+        else:
+            return response_upload_image
+    elif post.text:
+        response = requests.post(api_url_ugc, headers=headers, json=post_data)
+        return response
+
+
+def appropriate_response_action(response):
+    """Function to perform appropriate action based on the response given.
+
+    :param response: A valid HTTP response to perform suitable action.
+
+    :raises WrongArguments: Exception raised when either a 400 or 404 response is provided.
+
+    :raises NotAuthenticated: Exception raised when a 401 response is provided.
+
+    :raises PermissionDenied: Exception raised when a 403 response is provided.
+
+    :raises NotSupported: Exception raised when a 405 response is provided.
+
+    """
+    status_code = response.status_code
+    if status_code != status.HTTP_201_CREATED:
+        json_response = response.json()
+        error = f"LinkedIn error: {json_response['message']}"
+        if status_code == status.HTTP_400_BAD_REQUEST:
+            raise exc.WrongArguments(error)
+        elif status_code == status.HTTP_401_UNAUTHORIZED:
+            raise exc.NotAuthenticated(error)
+        elif status_code == status.HTTP_403_FORBIDDEN:
+            raise exc.PermissionDenied(error)
+        elif status_code == status.HTTP_404_NOT_FOUND:
+            raise exc.WrongArguments(error)
+        elif status_code == status.HTTP_405_METHOD_NOT_ALLOWED:
+            raise exc.NotSupported(error)
+
+
+def publish_post_service():
+    """Function to publish social media post"""
+    posts = Post.objects.filter(
+        is_approved=True, is_posted=False, scheduled_time__lte=timezone.now()
+    )
+    for post in posts:
+        post.posted_time = timezone.now()
+        post.is_posted = True
+        post.save()
+    for post in posts:
+        if post.posted_at == 'linkedin':
+            response = post_to_linkedin(post.id)
```

Comment:
You don't wait from LinkedIn's response to return response from your API.

Please use celery here.

Reply:
This service is being called within the task run by celery. So, isn't it already using celery?

Reply:
@gutsytechster currently, your one celery task that is publishing all the posts. Suppose, if the task gets failed, then we'll never know which `post` actually failed the task. Moreover, in this condition failing to publish one `post` will stop publishing rest of the posts as well.

A better approach is to make this task to call separate celery task for each post. This way, if a task will fail, we'll know which post is actually failing to publish.